File: public/code/hybrid__ema_adx.ts
import { Candle, Signal, makeSignal, sma, ema, rsi, macd, atr, bollingerBands } from "../indicators";

const mkH = (c: Candle[], i: number, side: "long" | "short", a: number[], conf: number, reason: string, m = 2): Signal => {
  const cur = c[i];
  if (side === "long") { const sl = cur.close - m * a[i], r = cur.close - sl; return makeSignal({ signal: "long", entry: cur.close, stop_loss: sl, take_profit: [cur.close + r * 1.5, cur.close + r * 2.5, cur.close + r * 4], confidence: conf, reason }); }
  const sl = cur.close + m * a[i], r = sl - cur.close; return makeSignal({ signal: "short", entry: cur.close, stop_loss: sl, take_profit: [cur.close - r * 1.5, cur.close - r * 2.5, cur.close - r * 4], confidence: conf, reason });
};
function stochK(c: Candle[], p = 14) {
  const out: number[] = [];
  for (let i = 0; i < c.length; i++) {
    if (i < p - 1) { out.push(50); continue; }
    const w = c.slice(i - p + 1, i + 1), hh = Math.max(...w.map((x) => x.high)), ll = Math.min(...w.map((x) => x.low));
    out.push(hh === ll ? 50 : ((c[i].close - ll) / (hh - ll)) * 100);
  }
  return sma(out, 3);
}

export function emaAdx(c: Candle[]): Signal {
  if (c.length < 40) return makeSignal({ reason: "Insufficient data" });
  const closes = c.map((x) => x.close), e1 = ema(closes, 9), e2 = ema(closes, 21), i = c.length - 1, a = atr(c, 14);
  // ADX proxy
  const p = 14, pdm: number[] = [], mdm: number[] = [], tr: number[] = [];
  for (let k = 1; k < c.length; k++) { const up = c[k].high - c[k - 1].high, dn = c[k - 1].low - c[k].low; pdm.push(up > dn && up > 0 ? up : 0); mdm.push(dn > up && dn > 0 ? dn : 0); tr.push(Math.max(c[k].high - c[k].low, Math.abs(c[k].high - c[k - 1].close), Math.abs(c[k].low - c[k - 1].close))); }
  const trS = ema(tr, p), pdi = ema(pdm, p).map((v, k) => (trS[k] ? (100 * v) / trS[k] : 0)), mdi = ema(mdm, p).map((v, k) => (trS[k] ? (100 * v) / trS[k] : 0));
  const dx = pdi.map((v, k) => (v + mdi[k] ? (100 * Math.abs(v - mdi[k])) / (v + mdi[k]) : 0)), adx = ema(dx, p);
  const j = adx.length - 1;
  if (e1[i - 1] <= e2[i - 1] && e1[i] > e2[i] && adx[j] > 25) return mkH(c, i, "long", a, 0.74, `EMA cross up + ADX strong (${adx[j].toFixed(0)})`);
  if (e1[i - 1] >= e2[i - 1] && e1[i] < e2[i] && adx[j] > 25) return mkH(c, i, "short", a, 0.74, `EMA cross down + ADX strong (${adx[j].toFixed(0)})`);
  return makeSignal({ reason: "No EMA+ADX trend signal" });
}

File: public/code/hybrid__ema_adx.py
from common import make_signal, ema, sma, rsi, macd, atr, bollinger_bands, vwap, swing_highs, swing_lows
import math


def _sign(x):
    return (x > 0) - (x < 0)


def mkH(c, i, side, a, conf, reason, m=2):
    cur = c[i]
    if side == "long":
        sl = cur['close'] - m * a[i]
        r = cur['close'] - sl
        return make_signal(signal="long", entry=cur['close'], stop_loss=sl, take_profit=[cur['close'] + r * 1.5, cur['close'] + r * 2.5, cur['close'] + r * 4], confidence=conf, reason=reason)
    sl = cur['close'] + m * a[i]
    r = sl - cur['close']
    return make_signal(signal="short", entry=cur['close'], stop_loss=sl, take_profit=[cur['close'] - r * 1.5, cur['close'] - r * 2.5, cur['close'] - r * 4], confidence=conf, reason=reason)
def stochK(c, p=14):
    out = []
    for i in range(0, len(c)):
        if i < p - 1:
            out.append(50)
            continue
        w = c[i - p + 1: i + 1]
        hh = max([x['high'] for x in w])
        ll = min([x['low'] for x in w])
        out.append((50 if hh == ll else ((c[i]['close'] - ll) / (hh - ll)) * 100))
    return sma(out, 3)
def emaAdx(c):
    if len(c) < 40:
        return make_signal(reason="Insufficient data")
    closes = [x['close'] for x in c]
    e1 = ema(closes, 9)
    e2 = ema(closes, 21)
    i = len(c) - 1
    a = atr(c, 14)
    # ADX proxy
    p = 14
    pdm = []
    mdm = []
    tr = []
    for k in range(1, len(c)):
        up = c[k]['high'] - c[k - 1]['high']
        dn = c[k - 1]['low'] - c[k]['low']
        pdm.append((up if up > dn  and  up > 0 else 0))
        mdm.append((dn if dn > up  and  dn > 0 else 0))
        tr.append(max(c[k]['high'] - c[k]['low'], abs(c[k]['high'] - c[k - 1]['close']), abs(c[k]['low'] - c[k - 1]['close'])))
    trS = ema(tr, p)
    pdi = [(((100 * v) / trS[k] if trS[k] else 0)) for k, v in enumerate(ema(pdm, p))]
    mdi = [(((100 * v) / trS[k] if trS[k] else 0)) for k, v in enumerate(ema(mdm, p))]
    dx = [(((100 * abs(v - mdi[k])) / (v + mdi[k]) if v + mdi[k] else 0)) for k, v in enumerate(pdi)]
    adx = ema(dx, p)
    j = len(adx) - 1
    if e1[i - 1] <= e2[i - 1]  and  e1[i] > e2[i]  and  adx[j] > 25:
        return mkH(c, i, "long", a, 0.74, f"EMA cross up + ADX strong ({adx[j]})")
    if e1[i - 1] >= e2[i - 1]  and  e1[i] < e2[i]  and  adx[j] > 25:
        return mkH(c, i, "short", a, 0.74, f"EMA cross down + ADX strong ({adx[j]})")
    return make_signal(reason="No EMA+ADX trend signal")
